From 7fb84473c4fca59c2d192b27c276dc8692eb810b Mon Sep 17 00:00:00 2001
From: Sudhakar Krishnan <sudhakar25@gmail.com>
Date: Sun, 8 Nov 2015 18:10:26 -0800
Subject: [PATCH] sdhci h file DMA latency

---
 drivers/mmc/host/sdhci.h | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/mmc/host/sdhci.h b/drivers/mmc/host/sdhci.h
index 32913e0..0f2829d 100644
--- a/drivers/mmc/host/sdhci.h
+++ b/drivers/mmc/host/sdhci.h
@@ -309,6 +309,12 @@ struct sdhci_adma2_64_desc {
  */
 #define SDHCI_MAX_SEGS		128
 
+enum sdhci_cookie {
+         COOKIE_UNMAPPED,
+         COOKIE_MAPPED,
+         COOKIE_GIVEN,
+};
+
 struct sdhci_host_next {
 	unsigned int	sg_count;
 	s32		cookie;
-- 
1.9.1

